src/algebra/category/Module/change_of_rings.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2022 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import algebra.category.Module.basic
+import algebra.category.Algebra.basic
+import ring_theory.tensor_product
+
+/-!
+# Change Of Rings
+
+## Main definitions
+
+* `category_theory.Module.restrict_scalars`: given rings `R, S` and a ring homomorphism `R ⟶ S`,
+  then `restrict_scalars : Module S ⥤ Module R` is defined by `M ↦ M` where `M : S-module` is seen
+  as `R-module` by `r • m := f r • m` and `S`-linear map `l : M ⟶ M'` is `R`-linear as well.
+
+## List of notations
+Let `R, S` be rings and `f : R →+* S`
+* if `M` is an `S`-module, `r : R` and `m : M` then notation `r r•[f] m` means `R`-scalar action on
+  `M` defined by `f r • m`.
+-/
+
+namespace category_theory.Module
+
+universes v u₁ u₂
+
+namespace restrict_scalars
+
+variables {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S)
+
+section unbundled
+
+variables (M : Type v) [add_comm_monoid M] [module S M]
+
+/-- The `R`-scalar multiplication on `S`-module M defined by `r • m := f r • m` -/
+protected def has_smul : has_smul R M :=
+(module.comp_hom M f).to_has_smul
+
+localized "notation r ` r•[` f `] ` :=
+  @@has_smul.smul (restrict_scalars.has_smul f _) r" in change_of_rings
+
+@[simp] lemma smul_def (r : R) (m : M) : r r•[f] m = f r • m := rfl
+
+end unbundled
+
+open_locale change_of_rings
+
+variable (M : Module.{v} S)
+
+/-- Any `S`-module M is also an `R`-module via a ring homomorphism `f : R ⟶ S` by defining
+    `r • m := f r • m` (`module.comp_hom`). This is called restriction of scalars. -/
+def obj' : Module R :=
+{ carrier := M,
+  is_add_comm_group := infer_instance,
+  is_module := module.comp_hom M f }
+
+/--
+Given an `S`-linear map `g : M → M'` between `S`-modules, `g` is also `R`-linear between `M` and
+`M'` by means of restriction of scalars.
+-/
+@[simps] def map' {M M' : Module.{v} S} (g : M ⟶ M') :
+  obj' f M ⟶ obj' f M' :=
+{ map_smul' := λ r (x : M), by simp,
+  ..g }
+
+end restrict_scalars
+
+/--
+ The restriction of scalars operation is functorial. For any `f : R →+* S` a ring homomorphism,
+ * an `S`-module `M` can be considered as `R`-module by `r • m = f r • m`
+ * an `S`-linear map is also `R`-linear
+ -/
+@[simps] def restrict_scalars {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S) :
```

Comment:
Is it true that `(Module.restrict_scalars f).obj M` is reduced by simp to `restrict_scalars.obj' f M`? It this something we want?

Relatedly, if `m : (Module.restrict_scalars f).obj M`, is `r \smul m` simplified to `f r \smul m`?

Reply:
> Is it true that (Module.restrict_scalars f).obj M is reduced by simp to restrict_scalars.obj' f M? It this something we want?
Yes, but this reduction isn't really useful. But useful lemmas are `restrict_scalars_map`, with this lemma the following is provable by simp
```lean
example {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S) (M M' : Module.{v} S)
  (g : M ⟶ M')
  (m : (restrict_scalars f).obj M) : (restrict_scalars f).map g m = g m :=
```
(actually left hand side and right hand side are `rfl`, but `simp` can prove this)

> Relatedly, if m : (Module.restrict_scalars f).obj M, is r \smul m simplified to f r \smul m?
This is provable by `rfl`
```lean
example {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S) (M : Module.{v} S) (m : (restrict_scalars f).obj M) (r : R) : r • m = f r • m := rfl
```

Reply:
Hmm, I think that we should make sure that we have a sensible `dsimp` normal form here. As I understand it, you intended `restrict_scalars.obj'` as an implementation detail rather than part of the "official" API. If this is true, I think it would be best to remove the `simps` attribute here and provide all relevant lemmas by hand.

Reply:
I think the `dsimp` normal form should be `(f r • m : M)`. I have removed all auto generated simp lemmas and wrote these two
```lean
@[simp] lemma restrict_scalars.map_apply {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S)
  {M M' : Module.{v} S} (g : M ⟶ M') (x) : (restrict_scalars f).map g x = g x := rfl

@[simp] lemma restrict_scalars.smul_def {R : Type u₁} {S : Type u₂} [ring R] [ring S] (f : R →+* S)
  {M : Module.{v} S} (r : R) (m : (restrict_scalars f).obj M) : r • m = (f r • m : M) := rfl
```
Are there any other lemmas that you think might be useful?

Reply:
These look good, if we find that more lemmas are needed down the line we can still add them later.